### PR TITLE
`struct Dav1dFrameHeader`: backport memory size reduction from dav1d 1.3.0

### PR DESCRIPTION
**Pull request by fbossen**
* Fixes `Dav1dFrameHeader` of #505.

Shrinks `Rav1dFrameHeader` from 1664 to 1120 bytes.

### CODE REVIEW

File: src/obu.rs [hunk 1]
```diff
@@ -704,8 +704,8 @@ unsafe fn parse_refidx(
                         .frame_hdr
                         .as_ref()
                         .ok_or(EINVAL)?
-                        .frame_offset,
-                    frame_offset,
+                        .frame_offset as c_int,
+                    frame_offset as c_int,
```
Comment (kkysen) [inline src/obu.rs:708]: We could make `fn get_poc_diff` and all of the `poc` fields/vars `u8`s.  I can add that if you want (I checked to see if it it's possible).

**Review by kkysen — CHANGES_REQUESTED** [review]
Mostly good, I just have some small comments about casts.  And I think `sharpness` is supposed to be a `u8` like in C.

File: include/dav1d/headers.rs [hunk 2]
```diff
@@ -2276,25 +2276,25 @@ impl From<Rav1dFrameHeader_delta> for Dav1dFrameHeader_delta {
...
-    pub mode_ref_delta_enabled: c_int,
-    pub mode_ref_delta_update: c_int,
+    pub level_y: [u8; 2],
+    pub level_u: u8,
+    pub level_v: u8,
+    pub mode_ref_delta_enabled: u8,
+    pub mode_ref_delta_update: u8,
     pub mode_ref_deltas: Rav1dLoopfilterModeRefDeltas,
-    pub sharpness: c_int,
+    pub sharpness: i8,
```
[the remark this answers is not in the document]
Comment (fbossen) [reply include/dav1d/headers.rs:2297]: To simplify comparing/assigning to `last_sharpness` which is `i8`

File: src/obu.rs [hunk 3]
```diff
@@ -900,21 +898,21 @@ fn parse_tiling(
...
+        log2_rows = tile_log2(1, rows.into());
     }
     col_start_sb[cols as usize] = sbw as u16;
     row_start_sb[rows as usize] = sbh as u16;
     let update;
     let n_bytes;
     if log2_cols != 0 || log2_rows != 0 {
-        update = gb.get_bits(log2_cols + log2_rows) as c_int;
-        if update >= cols * rows {
+        update = gb.get_bits((log2_cols + log2_rows).into()) as u16;
```
[the remark this answers is not in the document]
Comment (fbossen) [reply src/obu.rs:908]: Yes. `log_2_{cols,rows}` should be `<= 6` because `RAV1D_MAX_TILE_{COLS,ROWS}` is 64.

File: src/obu.rs (hunk 1 above)
Comment (fbossen) [reply src/obu.rs:708]: Yes, but that's really beyond the scope of this PR. I'll let you create a new one.

File: include/dav1d/headers.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkysen) [reply include/dav1d/headers.rs]: Hmm, maybe it should be an `Option`.  This is good for now, though.

**Review by kkysen — APPROVED** [review]

File: src/decode.rs [hunk 4]
```diff
@@ -358,7 +359,7 @@ unsafe fn read_tx_tree(
     };
 }
 
-fn neg_deinterleave(diff: c_int, r#ref: c_int, max: c_int) -> c_int {
+fn neg_deinterleave(diff: u8, r#ref: u8, max: u8) -> u8 {
```
Comment (rinon) [inline src/decode.rs:362]: We can't change these to unsigned, `max - 1` may be negative.

Comment (fbossen) [reply src/decode.rs:362]: Does #1056 solve it?

**Pull request merged**